### PR TITLE
Implement backend for favorite item feature by adding FavoriteItemServlet and FavoriteItemDatastore

### PR DESCRIPTION
**Pull request by BryanTriana**

### CODE REVIEW

File: src/test/java/com/google/ehub/data/FavoriteItemDatastoreTest.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    helper.setUp();
+  }
+
+  @After
+  public void tearDown() {
+    helper.tearDown();
+  }
+
+  @Test
+  public void addFavoriteItemToDatastore_EntityGetsAddedWithValidKindAndProperties() {
```
Comment (rharagutchi) [inline src/test/java/com/google/ehub/data/FavoriteItemDatastoreTest.java:45]: nit: after underscore, start again with lowercase

addFavoriteItemToDatastore_entityGetsAddedWithValidKindAndProperties

File: src/test/java/com/google/ehub/data/FavoriteItemDatastoreTest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    Assert.assertTrue(actualIds.containsAll(favoriteIds));
+  }
+
+  @Test
+  public void queryEmailsWithNonExistemItemId_ListIsEmpty() {
+    Assert.assertTrue(favoriteItemDatastore.queryEmails(ITEM_ID).isEmpty());
+  }
+
+  @Test
+  public void queryEmailsWithItemWithLikes_ListContainsEmailsThatLiked() {
```
Comment (rharagutchi) [inline src/test/java/com/google/ehub/data/FavoriteItemDatastoreTest.java:83]: add a test to make sure that duplicate favorites for the same person can't be added

File: src/main/java/com/google/ehub/servlets/FavoriteItemServlet.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+    if (!doesItemExist(itemId)) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST,
+          "FavoriteItemServlet: Favorite Item Id doesn't map to a known item!");
+      return;
+    }
+
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (!userService.isUserLoggedIn()) {
```
Comment (rharagutchi) [inline src/main/java/com/google/ehub/servlets/FavoriteItemServlet.java:55]: Does userLoggedIn imply user has a profile? A little worried you'll be adding favorites for people who don't actually have profiles which means you won't be able to recommend friend profiles, just the friend email. (not a huge deal, but recommending profiles would be nice, can be done as separate PR)

File: src/main/java/com/google/ehub/data/FavoriteItemDatastore.java [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+   *
+   * @param userEmail The email of the user that liked the item
+   * @param itemId the id of the item that was liked
+   */
+  public void addFavoriteItem(String userEmail, Long itemId) {
+    Entity favoriteItemEntity = new Entity(FAVORITE_ITEM_KIND);
+    favoriteItemEntity.setProperty(USER_EMAIL_PROPERTY_KEY, userEmail);
+    favoriteItemEntity.setProperty(ITEM_ID_PROPERTY_KEY, itemId);
+
+    datastoreService.put(favoriteItemEntity);
```
[the remark this answers is not in the document]
Comment (BryanTriana) [reply src/main/java/com/google/ehub/data/FavoriteItemDatastore.java:57]: Yeah I forgot about checking for that. I added the check in last commit.

File: src/test/java/com/google/ehub/data/FavoriteItemDatastoreTest.java (hunk 1 above)
Comment (BryanTriana) [reply src/test/java/com/google/ehub/data/FavoriteItemDatastoreTest.java]: I renamed the test method calls to use lowercase after underscore in last commit.

File: src/test/java/com/google/ehub/data/FavoriteItemDatastoreTest.java (hunk 2 above)
Comment (BryanTriana) [reply src/test/java/com/google/ehub/data/FavoriteItemDatastoreTest.java]: I added a test to check for duplicate items in last commit.

File: src/main/java/com/google/ehub/servlets/FavoriteItemServlet.java (hunk 3 above)
Comment (BryanTriana) [reply src/main/java/com/google/ehub/servlets/FavoriteItemServlet.java:58]: I think the way that it is setup right now it is not really asking if the user has a profile. I am going to ask Oyin if there is a way to know if the user has created the profile and it's logged in at the same time.

File: src/main/java/com/google/ehub/data/FavoriteItemDatastore.java [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+  public void addFavoriteItem(String userEmail, Long itemId) {
+    if (doesFavoriteItemExist(userEmail, itemId)) {
+      return;
+    }
+
+    Entity favoriteItemEntity = new Entity(FAVORITE_ITEM_KIND);
+    favoriteItemEntity.setProperty(USER_EMAIL_PROPERTY_KEY, userEmail);
+    favoriteItemEntity.setProperty(ITEM_ID_PROPERTY_KEY, itemId);
+
+    datastoreService.put(favoriteItemEntity);
```
Comment (codewithjess) [inline src/main/java/com/google/ehub/data/FavoriteItemDatastore.java:57]: Typically when you are accessing a database multiple times (e.g. read, then write) as a single action, you should use a transaction. This prevents the item from being modified by another user/process in between your calls to getItem() and put(). 

Here is a good example: https://cloud.google.com/appengine/docs/standard/java/datastore/transactions#uses_for_transactions

I know we haven't really discussed transactions that much but it helps prevent some race conditions that might be hard to detect. Feel free to add an issue for this and resolve in a later PR.

File: src/main/java/com/google/ehub/servlets/FavoriteItemServlet.java (hunk 3 above)
Comment (codewithjess) [reply src/main/java/com/google/ehub/servlets/FavoriteItemServlet.java:58]: Sounds good. Another possibility is to force the user to create a profile the first time they login. So all users are guaranteed to have a profile

File: src/main/java/com/google/ehub/data/FavoriteItemDatastore.java (hunk 5 above)
Comment (BryanTriana) [reply src/main/java/com/google/ehub/data/FavoriteItemDatastore.java:57]: OK I created a new issue for this.
